katharsis-core/src/main/java/io/katharsis/core/internal/utils/MethodCache.java

```diff
@@ -0,0 +1,64 @@
+package io.katharsis.core.internal.utils;
+
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.apache.commons.lang3.ObjectUtils;
+
+import io.katharsis.utils.Optional;
+
+public class MethodCache {
+
+	private Map<MethodCacheKey, Optional<Method>> cache = new ConcurrentHashMap<>();
```

Comment:
Guava cache has a `get` method that allows you to supply a function of what to do on a miss.